pulp_ostree/app/tasks/uploading.py

```diff
@@ -0,0 +1,263 @@
+                    commit_dc = self.create_artifact_dc(relative_path, commit)
+                    commit_dcs.append(commit_dc)
+
+                    checksum = parent_checksum
+                    relative_path = get_checksum_filepath(
+                        checksum, OstreeObjectType.OSTREE_OBJECT_TYPE_COMMIT
+                    )
+                    _, parent_commit, _ = self.repo.load_commit(checksum)
+                    parent_checksum = OSTree.commit_get_parent(parent_commit)
+
+                commit = OstreeCommit(checksum=checksum)
+                commit_dc = self.create_artifact_dc(relative_path, commit)
+
+                commit_dcs.append(commit_dc)
+
+                await self.put(commit_dc)
+                await self.submit_related_objects(commit)
+
+                for i in range(len(commit_dcs) - 1):
+                    commit_dcs[i].extra_data["parent_commit"] = commit_dcs[i + 1]
+
+                    await self.put(commit_dcs[i])
+                    await self.submit_related_objects(commit_dcs[i].content)
+
+                refshead = OstreeRefsHead(name=name, commit=headrefs_commit_dc.content)
+                refshead_path = os.path.join("refs/", "heads/", name)
+                refshead_dc = self.create_artifact_dc(refshead_path, refshead)
+                await self.put(refshead_dc)
+
+            pb.increment()
+
+    def init_repository(self):
+        """Initialize a new OSTree repository object."""
+        self.repo_path = os.path.join(os.getcwd(), self.repository_name)
+        self.repo = OSTree.Repo.new(Gio.File.new_for_path(self.repo_path))
+
+        try:
+            self.repo.open()
+        except GLib.Error:
+            raise ValueError(
+                gettext("An invalid path to the repository provided: {}").format(
+                    self.repository_name
+                )
+            )
+
+    def create_artifact_dc(self, relative_file_path, content):
+        """Create a DeclarativeContent object describing a single OSTree object (e.g., commit)."""
+        filepath = os.path.join(self.repo_path, relative_file_path)
+
+        # this is a hack that prevents the pipeline to remove a temporary file which can be
+        # referenced by multiple content units at the same time; in particular, this means that
+        # one OSTree object (e.g., dirmeta, file) is referenced by two different commits
+        filepath_copy = filepath + str(uuid.uuid4())
+        os.link(filepath, filepath_copy)
+        artifact = Artifact.init_and_validate(filepath_copy)
+
+        publication_filepath = os.path.join(self.repository_name, relative_file_path)
+        # DeclarativeArtifact requires a URL to be passed to its constructor even though it will
+        # never be used; specifying the URL is a requirement for standard downloading pipeline that
+        # we are not utilizing right now
+        da = DeclarativeArtifact(
+            artifact=artifact,
+            url="hackathon",
```

Comment:
Not sure what you mean by this comment. When I tested the code, it correctly determined which artifacts are existing and which do not.

Reply:
If an artifact didn't exist, it would try to download it from "hackathon", no?

Reply:
Sadly, you are right. But, this would only happen when somebody removes the artifact and the referencing content unit will be still present. When parsing a tarball, we always have a referenced artifact in place. Is there any way of enforcing on_delete.CASCADE for content units after removing associated artifacts? Or how could I resolve this issue? Using `DeclarativeArtifact` allowed me to simply define the relative path for a content unit that will be published. And that was the main reason of using it.

Reply:
Well, I'm not saying this is the wrong approach. I wanted to understand it. And as your pipeline didn't even have the `DownloadArtifacts` stage, it would probably fail at `SaveArtifacts`. And we have safeguards in place, so that an Artifact will not be cleaned as orphan for at least a day. So you should be fine.

Reply:
Does the pipeline create `RemoteArtifact`s in it's current state? That would be a bad idea.

Reply:
I think it does not create `RemoteArtifact`. These types of objects are created in `RemoteArtifactSaver` which I do not use at all. Then, I am leaving this part of the code untouched for now. Are you fine with that?